docs/adr/0007-preventing-version-mismatches.md

```diff
@@ -0,0 +1,269 @@
+with `uniffi` if their major versions match.  We would update
+`ComponentInterface` so it only uses the major version number when calculating
+it's hash.
+
+Pros:
+  - Seems like a hard requirement if we want external bindings tools
+  - Would significantly reduce the chance of running into a version mismatch.
+  - Makes it easier to detect version mismatches when parsing `Cargo.toml`.
+    Right now we need to determine the exact version that was resolved and
+    stored in `Cargo.lock`.  If we only needed to compare the major versions,
+    then we could use the specification from `Cargo.toml` directly.
+
+Cons
+  - We would have to be more careful with breaking changes. If we realized a
+    change would break the generated code compatibility, then that change might
+    need to wait before releasing it. If we didn't realize a change would break
+    the generated code compatilibity and released the change anyway, then we've
+    introduced undefined behavior in all consumer code.
+  - Doesn't fully solve the issue of version mismatches. We would need to
+    combine this with at one or more options below.
+
+### **[Option 1] generate bindings in `build.rs`**
+
+As mentioned in the context, we handle this issue for scaffolding by using the
+`builtin-bindgen` feature of `uniffi_build`.  If we also generated the bindings
+code in `build.rs`, we could use a similar system:
+
+- The Rust crate adds a dependency to each bindings generator they want to
+  use (`uniffi-bindgen-kotlin = 1.10.1`, `uniffi-bindgen-swift = 1.4.0`,
+  etc).
+- Each bindings generator defines a `generate_bindings()` function, which
+  generates the bindings and writes them to the `target/[language]` directory.
+- We call this in `build.rs`:
+
+```
+fn main() {
+  uniffi_bindgen_language1.generate_bindings();
+  uniffi_bindgen_language2.generate_bindings();
+  ...
+}
+```
+
+- The build scripts would then copy the bindings from that directory to the
+  final location.
+
+Pros:
+  - Doesn't rely on a system-wide install of `uniffi-bindgen`.
+  - Bindgen tool versions are specified in the `Cargo.toml`, alongside the
+    `uniffi` version.
+
+Cons:
+  - Forces a specific order to the build process.  If you try building the
+    bindings before the Rust component, it would fail since the source files
+    haven't been generated yet (or worse, succeed with stale code).
+  - Hard to make this work with downstream crates.  For the desktop JS example,
+    the app-services components would need to include a call to
+    `uniffi_bindgen_gecko_js.generate_bindings()` in the `build.rs` file which
+    is not ideal.  Also, it would generate Kotlin/Swift bindings, which aren't
+    needed.  I think the solution would be to create a feature gate for each
+    bindings language that we could enable/disable, but this seems overly
+    complex.  Even this solution wouldn't work for some 3rd-party cases, for
+    example a 3rd-party that wants to use their home-grown bindings generator.
```

Comment:
I'm not convinced features would be bad fit here TBH. I suspect that we are going to end up feature-gating stuff based on the platform anyway (eg, sync code is likely to be different at the very least)

(to be clear, I'm not advocating for this as the accepted solution, but I think this as worded overstates this con)

Reply:
Agreed.  I reworded this section.  It's still listed as a con, but it's emphasized less.